enforcer/src/main/java/org/wso2/micro/gateway/enforcer/security/jwt/InternalAPIKeyAuthenticator.java

```diff
@@ -0,0 +1,298 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.micro.gateway.enforcer.security.jwt;
+
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import com.nimbusds.jwt.proc.BadJWTException;
+import com.nimbusds.jwt.proc.DefaultJWTClaimsVerifier;
+import net.minidev.json.JSONArray;
+import net.minidev.json.JSONObject;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.micro.gateway.enforcer.api.RequestContext;
+import org.wso2.micro.gateway.enforcer.common.CacheProvider;
+import org.wso2.micro.gateway.enforcer.constants.APIConstants;
+import org.wso2.micro.gateway.enforcer.constants.APISecurityConstants;
+import org.wso2.micro.gateway.enforcer.dto.JWTTokenPayloadInfo;
+import org.wso2.micro.gateway.enforcer.exception.APISecurityException;
+import org.wso2.micro.gateway.enforcer.exception.MGWException;
+import org.wso2.micro.gateway.enforcer.security.AuthenticationContext;
+import org.wso2.micro.gateway.enforcer.security.Authenticator;
+import org.wso2.micro.gateway.enforcer.security.jwt.validator.RevokedJWTDataHolder;
+import org.wso2.micro.gateway.enforcer.util.FilterUtils;
+
+import java.text.ParseException;
+
+/**
+ * Implements the authenticator interface to authenticate request using a Internal Key.
+ */
+public class InternalAPIKeyAuthenticator implements Authenticator {
+
+    private static final Log log = LogFactory.getLog(InternalAPIKeyAuthenticator.class);
+    private String securityParam;
+
+    public InternalAPIKeyAuthenticator(String securityParam) {
+        this.securityParam = securityParam;
+    }
+
+    @Override
+    public boolean canAuthenticate(RequestContext requestContext) {
+        String internalKey = requestContext.getHeaders().get(APIConstants.JwtTokenConstants.INTERNAL_KEY);
+        if (internalKey != null && internalKey.split("\\.").length == 3) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public AuthenticationContext authenticate(RequestContext requestContext) throws APISecurityException {
+        if (requestContext.getMathedAPI() != null) {
+            if (log.isDebugEnabled()) {
+                log.info("Internal Key Authentication initialized");
+            }
+
+            try {
+                // Extract internal from the request while removing it from the msg context.
+                String internalKey = extractInternalKey(requestContext);
+                if (StringUtils.isEmpty(internalKey)) {
+                    log.error("Internal Key does not perform at request");
```

Comment:
Shall we make this message more meaningful

Reply:
+1

Reply:
Fixed by https://github.com/wso2/product-microgateway/pull/1747/commits/18efbb14f8e3d8b39eac836bf490639e3c126d6f